File: VB/Controllers/HomeController.vb
Imports Microsoft.VisualBasic
Imports System.Web.Mvc
Imports E4425.Models
Imports DevExpress.Web.Mvc

Namespace E4425.Controllers
	Public Class HomeController
		Inherits Controller
		Public Function Index() As ActionResult
			Return View(DataProvider.GetCustomers())
		End Function
		Public Function GridViewPartial() As ActionResult
			Return PartialView(DataProvider.GetCustomers())
		End Function
		Public Function GridViewEditPartial(ByVal modifiedValues As MVCxGridViewBatchUpdateValues(Of Customer, Integer)) As ActionResult
			For Each customerInfo In modifiedValues.Insert
				If modifiedValues.IsValid(customerInfo) Then
					DataProvider.InsertCustomer(customerInfo)
				End If
			Next customerInfo
			For Each customerInfo In modifiedValues.Update
				If modifiedValues.IsValid(customerInfo) Then
					DataProvider.UpdateCustomer(customerInfo)
				End If
			Next customerInfo
			For Each customerID In modifiedValues.DeleteKeys
				DataProvider.DeleteCustomer(customerID)
			Next customerID
			Return PartialView("GridViewPartial", DataProvider.GetCustomers())
		End Function

        Public Function ComboBoxCityPartial() As ActionResult
            Dim countryID As Integer = If((Request.Params("CountryID") IsNot Nothing), Integer.Parse(Request.Params("CountryID")), -1)
            ViewData("cities") = DataProvider.GetCities(countryID)
            Return PartialView(DataProvider.GetCustomers())
        End Function
	End Class
End Namespace

File: CS/Controllers/HomeController.cs
using System.Web.Mvc;
using E4425.Models;
using DevExpress.Web.Mvc;
using System.Linq;
namespace E4425.Controllers
{
    public class HomeController : Controller
    {
        WorldCitiesEntities entity = new WorldCitiesEntities();
        public ActionResult Index()
        {
            return View(entity.Customers.ToList());
        }
        public ActionResult GridViewPartial()
        {
            return PartialView(entity.Customers.ToList());
        }
        public ActionResult GridViewEditPartial(MVCxGridViewBatchUpdateValues<Customer, int> modifiedValues)
        {
            foreach (var customerInfo in modifiedValues.Insert)
            {
                if (modifiedValues.IsValid(customerInfo))
                    entity.Customers.Add(customerInfo);
            }
            foreach (var customerInfo in modifiedValues.Update)
            {
                if (modifiedValues.IsValid(customerInfo))
                {
                    entity.Customers.Attach(customerInfo);
                    var entry = entity.Entry(customerInfo);
                    entry.Property(e => e.CityId).IsModified = true;
                    entry.Property(e => e.CountryId).IsModified = true;
                    entry.Property(e => e.CustomerName).IsModified = true;
                }
            }
            foreach (var customerID in modifiedValues.DeleteKeys)
            {
                entity.Customers.Remove(entity.Customers.Find(customerID));
            }
            // uncomment the next line to enable database updates
            // entity.SaveChanges();
            return PartialView("GridViewPartial", entity.Customers.ToList());
        }
        public ActionResult ComboBoxCountryPartial()
        {
            return GridViewExtension.GetComboBoxCallbackResult(ComboBoxPropertiesProvider.Current.CountryComboBoxProperties);
        }
        public ActionResult ComboBoxCityPartial()
        {
            return GridViewExtension.GetComboBoxCallbackResult(ComboBoxPropertiesProvider.Current.CityComboBoxProperties);
        }
    }

}